node_modules/oae-principals/lib/internal/dao.js

```diff
@@ -184,28 +216,71 @@ var updatePrincipal = module.exports.updatePrincipal = function(principalId, pro
-    var q = Cassandra.constructUpsertCQL('Principals', 'principalId', principalId, profileFields);
-    if (!q) {
-        return callback({'code': 500, 'msg': 'Unable to store profile fields'});
+        var queries = [];
+
+        // Update the principal record
+        queries.push(Cassandra.constructUpsertCQL('Principals', 'principalId', principalId, profileFields));
+
+        // If the user's email address needs to be updated, we remove the old one from the mapping
+        if (isEmailAddressUpdate) {
+            queries.push({'query': 'DELETE FROM "PrincipalsByEmail" WHERE "email" = ?', 'parameters': [oldEmail]});
+            queries.push(Cassandra.constructUpsertCQL('PrincipalsByEmail', 'email', profileFields.email, {'principalId': principalId}));
+        }
+
+        // Execute the queries
+        Cassandra.runBatchQuery(queries, function(err) {
+            if (err) {
+                return callback(err);
+            }
+
+            if (isUser(principalId)) {
+                // Update the user in cache. Ensure the `principalId` is not
+                // part of the hash to avoid revalidating a stale cache entry
+                profileFields = _.extend({}, profileFields);
+                delete profileFields.principalId;
+                return _updateCachedUser(principalId, profileFields, callback);
+            } else {
+                return callback();
+            }
+        });
+    });
+};
+
+/**
+ * Check whether the email address for a user would be updated when updating a set of `profileFields`
+ *
+ * @param  {String}     principalId                         The id of the principal to update
+ * @param  {Object}     profileFields                       An object containing the profile field updates to apply
+ * @param  {Function}   callback                            Standard callback function
+ * @param  {Object}     callback.err                        An error that occurred, if any
+ * @param  {Boolean}    calllback.isEmailAddressUpdate      Whether the email address will be updated
+ * @param  {String}     callback.oldEmail                   The user's old email address
+ * @api private
+ */
+var _isEmailAddressUpdate = function(principalId, profileFields, callback) {
+    if (!isUser(principalId) || !profileFields.email) {
+        return callback(null, false);
     }
 
-    Cassandra.runQuery(q.query, q.parameters, function(err) {
+    getPrincipal(principalId, function(err, user) {
         if (err) {
             return callback(err);
         }
 
-        if (isUser(principalId)) {
-            // Update the user in cache
-
-            // Ensure the `principalId` is not part of the hash to avoid revalidating a stale cache entry
-            profileFields = _.extend({}, profileFields);
-            delete profileFields.principalId;
-            return _updateCachedUser(principalId, profileFields, callback);
-        } else {
-            return callback();
+        if (user.email !== profileFields.email) {
```

Comment:
Not a huge deal, but this might be easier to read in an `else if` since it's a simple one-liner break like the error callback